### PR TITLE
feat(ios): Added new a11y properties for managing font scale

### PR DESCRIPTION
**Pull request by CatchABus**
## PR Checklist

- [x] The PR title follows our guidelines: https://github.com/NativeScript/NativeScript/blob/master/tools/notes/CONTRIBUTING.md#commit-messages.
- [ ] There is an issue for the bug/feature this PR is for. To avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it.
- [x] You have signed the [CLA](http://www.nativescript.org/cla).
- [x] All existing tests are passing: https://github.com/NativeScript/NativeScript/blob/master/tools/notes/DevelopmentWorkflow.md#running-unit-tests-application.
- [x] Tests for the changes are included - https://github.com/NativeScript/NativeScript/blob/master/tools/notes/WritingUnitTests.md.

## What is the current behavior?
As of NativeScript 8.5, we added missing support for scaling font size on iOS using accessibility settings by default.
This caused problems for a number of developers as they don't want this feature or want to put limits to it.

## What is the new behavior?
This PR adds the following new properties for iOS:

- iOSAccessibilityAdjustsFontSize (default: false)
- iOSAccessibilityMinFontScale
- iOSAccessibilityMaxFontScale

All of them are inherited css properties and inline properties at the same time.
Assigned one to root view is enough for affecting entire view tree.

BREAKING CHANGES:

This PR ensures accessibility font scale is disabled by default on iOS.
It has minor to almost no impact as scaling was not available before 8.5.

### CODE REVIEW

**Review by NathanWalker — APPROVED** [review]

File: packages/core/accessibility/accessibility-properties.ts [hunk 1]
```diff
@@ -31,6 +31,30 @@ export const accessibilityEnabledProperty = new CssProperty<Style, boolean>({
 });
 accessibilityEnabledProperty.register(Style);
 
+export const iOSAccessibilityAdjustsFontSizeProperty = new InheritedCssProperty<Style, boolean>({
+	defaultValue: false,
+	name: 'iOSAccessibilityAdjustsFontSize',
```
Comment (NathanWalker) [inline packages/core/accessibility/accessibility-properties.ts:36]:
```suggestion
	name: 'iosAccessibilityAdjustsFontSize',
```

File: packages/core/accessibility/accessibility-properties.ts [hunk 2]
```diff
@@ -31,6 +31,30 @@ export const accessibilityEnabledProperty = new CssProperty<Style, boolean>({
...
+	cssName: 'ios-a11y-adjusts-font-size',
+	valueConverter: booleanConverter,
+});
+iOSAccessibilityAdjustsFontSizeProperty.register(Style);
+
+export const iOSAccessibilityMinFontScaleProperty = new InheritedCssProperty<Style, number>({
+	defaultValue: 0,
+	name: 'iOSAccessibilityMinFontScale',
```
Comment (NathanWalker) [inline packages/core/accessibility/accessibility-properties.ts:44]:
```suggestion
	name: 'iosAccessibilityMinFontScale',
```

File: packages/core/accessibility/accessibility-properties.ts [hunk 3]
```diff
@@ -31,6 +31,30 @@ export const accessibilityEnabledProperty = new CssProperty<Style, boolean>({
...
+	cssName: 'ios-a11y-min-font-scale',
+	valueConverter: parseFloat,
+});
+iOSAccessibilityMinFontScaleProperty.register(Style);
+
+export const iOSAccessibilityMaxFontScaleProperty = new InheritedCssProperty<Style, number>({
+	defaultValue: 0,
+	name: 'iOSAccessibilityMaxFontScale',
```
Comment (NathanWalker) [inline packages/core/accessibility/accessibility-properties.ts:52]:
```suggestion
	name: 'iosAccessibilityMaxFontScale',
```

File: packages/core/accessibility/accessibility-properties.ts [hunk 4]
```diff
@@ -31,6 +31,30 @@ export const accessibilityEnabledProperty = new CssProperty<Style, boolean>({
...
+	cssName: 'ios-a11y-max-font-scale',
+	valueConverter: parseFloat,
+});
+iOSAccessibilityMaxFontScaleProperty.register(Style);
```
Comment (NathanWalker) [inline packages/core/accessibility/accessibility-properties.ts:56]:
```suggestion
iosAccessibilityMaxFontScaleProperty.register(Style);
```

File: packages/core/accessibility/accessibility-properties.ts [hunk 5]
```diff
@@ -31,6 +31,30 @@ export const accessibilityEnabledProperty = new CssProperty<Style, boolean>({
...
+
+export const iOSAccessibilityMinFontScaleProperty = new InheritedCssProperty<Style, number>({
+	defaultValue: 0,
+	name: 'iOSAccessibilityMinFontScale',
+	cssName: 'ios-a11y-min-font-scale',
+	valueConverter: parseFloat,
+});
+iOSAccessibilityMinFontScaleProperty.register(Style);
+
+export const iOSAccessibilityMaxFontScaleProperty = new InheritedCssProperty<Style, number>({
```
Comment (NathanWalker) [inline packages/core/accessibility/accessibility-properties.ts:50]:
```suggestion
export const iosAccessibilityMaxFontScaleProperty = new InheritedCssProperty<Style, number>({
```

File: packages/core/accessibility/accessibility-properties.ts [hunk 6]
```diff
@@ -31,6 +31,30 @@ export const accessibilityEnabledProperty = new CssProperty<Style, boolean>({
...
+});
+iOSAccessibilityAdjustsFontSizeProperty.register(Style);
+
+export const iOSAccessibilityMinFontScaleProperty = new InheritedCssProperty<Style, number>({
+	defaultValue: 0,
+	name: 'iOSAccessibilityMinFontScale',
+	cssName: 'ios-a11y-min-font-scale',
+	valueConverter: parseFloat,
+});
+iOSAccessibilityMinFontScaleProperty.register(Style);
```
Comment (NathanWalker) [inline packages/core/accessibility/accessibility-properties.ts:48]:
```suggestion
iosAccessibilityMinFontScaleProperty.register(Style);
```

File: packages/core/accessibility/accessibility-properties.ts [hunk 7]
```diff
@@ -31,6 +31,30 @@ export const accessibilityEnabledProperty = new CssProperty<Style, boolean>({
...
 
+export const iOSAccessibilityAdjustsFontSizeProperty = new InheritedCssProperty<Style, boolean>({
+	defaultValue: false,
+	name: 'iOSAccessibilityAdjustsFontSize',
+	cssName: 'ios-a11y-adjusts-font-size',
+	valueConverter: booleanConverter,
+});
+iOSAccessibilityAdjustsFontSizeProperty.register(Style);
+
+export const iOSAccessibilityMinFontScaleProperty = new InheritedCssProperty<Style, number>({
```
Comment (NathanWalker) [inline packages/core/accessibility/accessibility-properties.ts:42]:
```suggestion
export const iosAccessibilityMinFontScaleProperty = new InheritedCssProperty<Style, number>({
```

File: packages/core/accessibility/accessibility-properties.ts [hunk 8]
```diff
@@ -31,6 +31,30 @@ export const accessibilityEnabledProperty = new CssProperty<Style, boolean>({
 });
 accessibilityEnabledProperty.register(Style);
 
+export const iOSAccessibilityAdjustsFontSizeProperty = new InheritedCssProperty<Style, boolean>({
+	defaultValue: false,
+	name: 'iOSAccessibilityAdjustsFontSize',
+	cssName: 'ios-a11y-adjusts-font-size',
+	valueConverter: booleanConverter,
+});
+iOSAccessibilityAdjustsFontSizeProperty.register(Style);
```
Comment (NathanWalker) [inline packages/core/accessibility/accessibility-properties.ts:40]:
```suggestion
iosAccessibilityAdjustsFontSizeProperty.register(Style);
```

File: packages/core/accessibility/accessibility-properties.ts [hunk 9]
```diff
@@ -31,6 +31,30 @@ export const accessibilityEnabledProperty = new CssProperty<Style, boolean>({
 });
 accessibilityEnabledProperty.register(Style);
 
+export const iOSAccessibilityAdjustsFontSizeProperty = new InheritedCssProperty<Style, boolean>({
```
Comment (NathanWalker) [inline packages/core/accessibility/accessibility-properties.ts:34]:
```suggestion
export const iosAccessibilityAdjustsFontSizeProperty = new InheritedCssProperty<Style, boolean>({
```

File: apps/automated/src/ui/styling/style-properties-tests.ts [hunk 10]
```diff
@@ -582,13 +582,17 @@ export function test_setting_font_properties_sets_native_font() {
...
 		const testView = new Label();
-		const deviceFontScaleMock = 4.0;
+		const layout = new StackLayout();
+		layout.addChild(testView);
 
-		page.content = testView;
+		page.content = layout;
 
-		testView.style._fontScale = deviceFontScaleMock;
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
```
Comment (NathanWalker) [inline apps/automated/src/ui/styling/style-properties-tests.ts:594]:
```suggestion
		layout.style.iosAccessibilityAdjustsFontSize = true;
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
+		layout.style.fontScaleInternal = deviceFontScaleMock;
+
+		testView.style.iOSAccessibilityMaxFontScale = 2.0;
+
+		const nativeFontSize = testView.nativeTextViewProtected.font.pointSize;
+		const expectedNativeFontSize = testView.style.fontInternal.fontSize * testView.style.iOSAccessibilityMaxFontScale;
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:67]:
```suggestion
		const expectedNativeFontSize = testView.style.fontInternal.fontSize * testView.style.iosAccessibilityMaxFontScale;
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+		const testView = new Label();
+		const layout = new StackLayout();
+		layout.addChild(testView);
+
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
+		layout.style.fontScaleInternal = deviceFontScaleMock;
+
+		testView.style.iOSAccessibilityMaxFontScale = 2.0;
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:64]:
```suggestion
		testView.style.iosAccessibilityMaxFontScale = 2.0;
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+		const deviceFontScaleMock = 4.0;
+
+		const page = helper.getCurrentPage();
+		const testView = new Label();
+		const layout = new StackLayout();
+		layout.addChild(testView);
+
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:61]:
```suggestion
		layout.style.iosAccessibilityAdjustsFontSize = true;
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+		testView.style.iOSAccessibilityMinFontScale = 2.0;
+
+		const nativeFontSize = testView.nativeTextViewProtected.font.pointSize;
+		const expectedNativeFontSize = testView.style.fontInternal.fontSize * testView.style.iOSAccessibilityMinFontScale;
+		TKUnit.assertEqual(nativeFontSize, expectedNativeFontSize, 'View font size scaling does not respect iOSAccessibilityMinFontScale');
+	}
+}
+
+export function test_iOSAccessibilityMaxFontScale_property() {
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:50]:
```suggestion
export function test_iosAccessibilityMaxFontScale_property() {
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
+		layout.style.fontScaleInternal = deviceFontScaleMock;
+
+		testView.style.iOSAccessibilityMinFontScale = 2.0;
+
+		const nativeFontSize = testView.nativeTextViewProtected.font.pointSize;
+		const expectedNativeFontSize = testView.style.fontInternal.fontSize * testView.style.iOSAccessibilityMinFontScale;
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:45]:
```suggestion
		const expectedNativeFontSize = testView.style.fontInternal.fontSize * testView.style.iosAccessibilityMinFontScale;
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+		const testView = new Label();
+		const layout = new StackLayout();
+		layout.addChild(testView);
+
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
+		layout.style.fontScaleInternal = deviceFontScaleMock;
+
+		testView.style.iOSAccessibilityMinFontScale = 2.0;
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:42]:
```suggestion
		testView.style.iosAccessibilityMinFontScale = 2.0;
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+		const deviceFontScaleMock = 1.0;
+
+		const page = helper.getCurrentPage();
+		const testView = new Label();
+		const layout = new StackLayout();
+		layout.addChild(testView);
+
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:39]:
```suggestion
		layout.style.iosAccessibilityAdjustsFontSize = true;
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+		const nativeFontSize = testView.nativeTextViewProtected.font.pointSize;
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
+		const nativeFontSizeWithAdjust = testView.nativeTextViewProtected.font.pointSize;
+
+		TKUnit.assertEqual(nativeFontSize, testView.style.fontInternal.fontSize, 'View font size was scaled even though iOSAccessibilityAdjustsFontSize is disabled');
+		TKUnit.assertEqual(nativeFontSizeWithAdjust, testView.style.fontInternal.fontSize * deviceFontScaleMock, 'View font size was not scaled even though iOSAccessibilityAdjustsFontSize is enabled');
+	}
+}
+
+export function test_iOSAccessibilityMinFontScale_property() {
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:28]:
```suggestion
export function test_iosAccessibilityMinFontScale_property() {
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+
+		layout.style.iOSAccessibilityAdjustsFontSize = false;
+		layout.style.fontScaleInternal = deviceFontScaleMock;
+
+		const nativeFontSize = testView.nativeTextViewProtected.font.pointSize;
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
+		const nativeFontSizeWithAdjust = testView.nativeTextViewProtected.font.pointSize;
+
+		TKUnit.assertEqual(nativeFontSize, testView.style.fontInternal.fontSize, 'View font size was scaled even though iOSAccessibilityAdjustsFontSize is disabled');
+		TKUnit.assertEqual(nativeFontSizeWithAdjust, testView.style.fontInternal.fontSize * deviceFontScaleMock, 'View font size was not scaled even though iOSAccessibilityAdjustsFontSize is enabled');
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:24]:
```suggestion
		TKUnit.assertEqual(nativeFontSizeWithAdjust, testView.style.fontInternal.fontSize * deviceFontScaleMock, 'View font size was not scaled even though iosAccessibilityAdjustsFontSize is enabled');
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = false;
+		layout.style.fontScaleInternal = deviceFontScaleMock;
+
+		const nativeFontSize = testView.nativeTextViewProtected.font.pointSize;
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
+		const nativeFontSizeWithAdjust = testView.nativeTextViewProtected.font.pointSize;
+
+		TKUnit.assertEqual(nativeFontSize, testView.style.fontInternal.fontSize, 'View font size was scaled even though iOSAccessibilityAdjustsFontSize is disabled');
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:23]:
```suggestion
		TKUnit.assertEqual(nativeFontSize, testView.style.fontInternal.fontSize, 'View font size was scaled even though iosAccessibilityAdjustsFontSize is disabled');
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+		const layout = new StackLayout();
+		layout.addChild(testView);
+
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = false;
+		layout.style.fontScaleInternal = deviceFontScaleMock;
+
+		const nativeFontSize = testView.nativeTextViewProtected.font.pointSize;
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:20]:
```suggestion
		layout.style.iosAccessibilityAdjustsFontSize = true;
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+		const deviceFontScaleMock = 4.0;
+
+		const page = helper.getCurrentPage();
+		const testView = new Label();
+		const layout = new StackLayout();
+		layout.addChild(testView);
+
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = false;
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:16]:
```suggestion
		layout.style.iosAccessibilityAdjustsFontSize = false;
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 23]
```diff
@@ -0,0 +1,70 @@
+import * as TKUnit from '../tk-unit';
+import * as helper from '../ui-helper';
+import { isIOS, Label, StackLayout } from '@nativescript/core';
+
+export function test_iOSAccessibilityAdjustsFontSize_property() {
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:5]:
```suggestion
export function test_iosAccessibilityAdjustsFontSize_property() {
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
+		layout.style.fontScaleInternal = deviceFontScaleMock;
+
+		testView.style.iOSAccessibilityMinFontScale = 2.0;
+
+		const nativeFontSize = testView.nativeTextViewProtected.font.pointSize;
+		const expectedNativeFontSize = testView.style.fontInternal.fontSize * testView.style.iOSAccessibilityMinFontScale;
+		TKUnit.assertEqual(nativeFontSize, expectedNativeFontSize, 'View font size scaling does not respect iOSAccessibilityMinFontScale');
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:46]:
```suggestion
		TKUnit.assertEqual(nativeFontSize, expectedNativeFontSize, 'View font size scaling does not respect iosAccessibilityMinFontScale');
```

File: apps/automated/src/accessibility/accessibility-properties-tests.ts [hunk 25]
```diff
@@ -0,0 +1,70 @@
...
+		page.content = layout;
+
+		layout.style.iOSAccessibilityAdjustsFontSize = true;
+		layout.style.fontScaleInternal = deviceFontScaleMock;
+
+		testView.style.iOSAccessibilityMaxFontScale = 2.0;
+
+		const nativeFontSize = testView.nativeTextViewProtected.font.pointSize;
+		const expectedNativeFontSize = testView.style.fontInternal.fontSize * testView.style.iOSAccessibilityMaxFontScale;
+		TKUnit.assertEqual(nativeFontSize, expectedNativeFontSize, 'View font size scaling does not respect iOSAccessibilityMaxFontScale');
```
Comment (NathanWalker) [inline apps/automated/src/accessibility/accessibility-properties-tests.ts:68]:
```suggestion
		TKUnit.assertEqual(nativeFontSize, expectedNativeFontSize, 'View font size scaling does not respect iosAccessibilityMaxFontScale');
```

File: packages/core/ui/core/view/index.d.ts [hunk 26]
```diff
@@ -282,6 +282,21 @@ export abstract class View extends ViewCommon {
 	 */
 	accessibilityMediaSession: boolean;
 
+	/**
+	 * Defines whether accessibility font scale should affect font size.
+	 */
+	iOSAccessibilityAdjustsFontSize: boolean;
```
Comment (NathanWalker) [inline packages/core/ui/core/view/index.d.ts:288]:
```suggestion
	iosAccessibilityAdjustsFontSize: boolean;
```

File: packages/core/ui/core/view/index.d.ts [hunk 27]
```diff
@@ -282,6 +282,21 @@ export abstract class View extends ViewCommon {
...
+
+	/**
+	 * Gets or sets the minimum accessibility font scale.
+	 */
+	iOSAccessibilityMinFontScale: number;
```
Comment (NathanWalker) [inline packages/core/ui/core/view/index.d.ts:293]:
```suggestion
	iosAccessibilityMinFontScale: number;
```

File: packages/core/ui/core/view/index.d.ts [hunk 28]
```diff
@@ -282,6 +282,21 @@ export abstract class View extends ViewCommon {
...
+
+	/**
+	 * Gets or sets the maximum accessibility font scale.
+	 */
+	iOSAccessibilityMaxFontScale: number;
```
Comment (NathanWalker) [inline packages/core/ui/core/view/index.d.ts:298]:
```suggestion
	iosAccessibilityMaxFontScale: number;
```

File: packages/core/ui/core/view/view-common.ts [hunk 29]
```diff
@@ -859,6 +859,27 @@ export abstract class ViewCommon extends ViewBase implements ViewDefinition {
 		this.style.accessibilityMediaSession = value;
 	}
 
+	get iOSAccessibilityAdjustsFontSize(): boolean {
```
Comment (NathanWalker) [inline packages/core/ui/core/view/view-common.ts:862]:
```suggestion
	get iosAccessibilityAdjustsFontSize(): boolean {
```

File: packages/core/ui/core/view/view-common.ts [hunk 30]
```diff
@@ -859,6 +859,27 @@ export abstract class ViewCommon extends ViewBase implements ViewDefinition {
...
+		return this.style.iOSAccessibilityAdjustsFontSize;
```
Comment (NathanWalker) [inline packages/core/ui/core/view/view-common.ts:863]:
```suggestion
		return this.style.iosAccessibilityAdjustsFontSize;
```

**Review by NathanWalker — CHANGES_REQUESTED** [review]
@CatchABus very nice - if you don't mind, let's stick with `ios` prefix on platform specific properties to match the approach of `iosOverflowSafeArea`.

**Comment by CatchABus** [comment]
> @CatchABus very nice - if you don't mind, let's stick with `ios` prefix on platform specific properties to match the approach of `iosOverflowSafeArea`.

My bad...thanks a lot for pointing out! I think all of them are corrected now.

**Review by NathanWalker — APPROVED** [review]

**Pull request merged**